internal/service/storage/storage_data_source.go

```diff
@@ -0,0 +1,232 @@
+
+Data source can be used to map storage to other resource based on the ID or just to read some other storage property like zone information. Storage types are: ` + "`" + `normal` + "`" + `, ` + "`" + `backup` + "`" + `, ` + "`" + `cdrom` + "`" + `, and ` + "`" + `template` + "`" + `.`
+
+func (d *storageDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: storageDataSourceDescription,
+		Attributes: map[string]schema.Attribute{
+			"access_type": schema.StringAttribute{
+				MarkdownDescription: "The access type of the storage, `public` or `private`.",
+				Computed:            true,
+				Optional:            true,
+				Validators: []validator.String{
+					stringvalidator.OneOf(
+						upcloud.StorageAccessPublic,
+						upcloud.StorageAccessPrivate,
+					),
+				},
+			},
+			"encrypt": schema.BoolAttribute{
+				MarkdownDescription: encryptDescription,
+				Computed:            true,
+			},
+			"id": schema.StringAttribute{
+				Computed:            true,
+				Optional:            true,
+				MarkdownDescription: uuidDescription,
+			},
+			"labels":        utils.ReadOnlyLabelsAttribute("storage"),
+			"system_labels": utils.SystemLabelsAttribute("storage"),
+			"size": schema.Int64Attribute{
+				MarkdownDescription: sizeDescription,
+				Computed:            true,
+			},
+			"state": schema.StringAttribute{
+				Description: "Current state of the storage",
+				Computed:    true,
+			},
+			"tier": schema.StringAttribute{
+				MarkdownDescription: tierDescription,
+				Computed:            true,
+			},
+			"title": schema.StringAttribute{
+				MarkdownDescription: titleDescription,
+				Optional:            true,
+				Computed:            true,
+			},
+			"type": schema.StringAttribute{
+				MarkdownDescription: typeDescription,
+				Computed:            true,
+				Optional:            true,
+			},
+			"zone": schema.StringAttribute{
+				MarkdownDescription: "The zone the storage is in, e.g. `de-fra1`.",
+				Optional:            true,
+				Computed:            true,
+			},
+			"name_regex": schema.StringAttribute{
+				Description:        "Use regular expression to match storage name. Deprecated, use exact title or UUID instead.",
+				DeprecationMessage: "Use exact title or UUID instead.",
+				Optional:           true,
+			},
+			"name": schema.StringAttribute{
+				Description:        "Exact name of the storage (same as title). Deprecated, use `title` instead.",
+				DeprecationMessage: "Contains the same value as `title`. Use `title` instead.",
+				Optional:           true,
+			},
+			"most_recent": schema.BoolAttribute{
+				Description:        "If more than one result is returned, use the most recent storage. This is only useful with private storages. Public storages might give unpredictable results.",
+				DeprecationMessage: "Use exact title or UUID to limit the number of matching storages to one.",
```

Comment:
Not a blocker, wondering if this could be re-phrased somehow as exact title might not help limiting the matches to 1, as it is not enforced to be unique.

Reply:
Maybe like this?

```suggestion
				DeprecationMessage: "Use exact title or UUID to limit the number of matching storages. Note that if you have multiple storages with the same title, you should use UUID to select the storage.",
```

Reply:
Nice